### DISCUSSION TITLE
How to catch authentication error's in davici initiate API?

### DISCUSSION THREAD

**Discussion by digambaringale333** [strongswan/strongswan#325] [Q&A]
Hi,
We are using davici code to implement IPSec in our application. We have couple of questions around initiate API.
1. Consider a case where I have wrong root and end point certificate installed in VICI, in this case if I issue initiate tunnel API, it completes without throwing  any errors (for eg: Authentication failure).
2. Is there any way, we can tell VICI that if all is good (auth-material is good then only send success else send failure)
3. If q-2 is not supported, then what is the way to know that the initiate API actually failed to created tunnel because of auth failure?
4. As per documentation initiate API may throw some error messages in some cases. Can you please point to us to see what are possible error messages initiate API may throw?
```
 {
    child = <CHILD_SA configuration name to initiate>
    ike = <IKE_SA configuration name to initiate or to find child under>
    timeout = <timeout in ms before returning>
    init-limits = <whether limits may prevent initiating the CHILD_SA>
    loglevel = <loglevel to issue "control-log" events for>
} => {
    success = <yes or no>
    errmsg = <error string on failure or timeout>
}

```
Thanks in advance.

**Comment by tobiasbrunner** [comment]
> Is there any way, we can tell VICI that if all is good (auth-material is good then only send success else send failure)

That should already be the case. An authentication failure should not result in `success = yes`. Unless you initiate with `timeout = -1`, as that will only initiate the connection and return immediately without waiting for the actual result of the negotiation.

> As per documentation initiate API may throw some error messages in some cases. Can you please point to us to see what are possible error messages initiate API may throw?

These are mostly errors if e.g. the config can't be found or parameters are incorrect. You won't get any details via vici if the connection can't be established. You could try parsing the log, but I wouldn't recommend that. However, using the [_error-notify_ plugin](https://wiki.strongswan.org/projects/strongswan/wiki/ErrorNotifyPlugin) might be an option to get slightly more detailed notifications (e.g. that the peer authentication failed, you still won't see why exactly). You can connect to the socket opened by the plugin from an external application.

**Reply by digambaringale333** [reply]
Hi Tobias, Thanks for your reply. I tried to use error-notify plugin as you mentioned, but I don't see any way to register a callback with  VICI to get VICI response for error-notify notifications(for commands and unsolicited events we can use davici_queue and davici_register to register our call back for vici response). Is there any way to get these error-notification messages  through callback?

**Reply by tobiasbrunner** [reply]
As documented (see the link above), the _error-notify_ plugin provides its own socket and protocol (have a look at the [demo client](https://github.com/strongswan/strongswan/blob/859dedeab751c71e55517a1f8d4761347e006a9d/src/libcharon/plugins/error_notify/error_notify.c)). So it has nothing to do with VICI. But you could read the socket from the same I/O loop (i.e. `poll()` call) as the VICI sockets.

**Reply by digambaringale333** [reply]
Ok.. Got it.. Thanks for your quick reply Tobias.
